### ISSUE TITLE
[Feature]: Allow oot custom compiler extension via CompilerInterface and reuse backend-agnostic FX passes

### ISSUE THREAD

**Issue by jgong5** [vllm-project/vllm#23612]
### 🚀 The feature, motivation and pitch

From vllm-ascend, we are trying to enable some graph fusion passes by extending the vllm core. But it seems that the compilation backend is currently coupled with the "inductor" backend which not every HW backend supports. In vllm-ascend, the inductor backend is not yet enabled and we are adding direct FX-based graph fusion. Currently, we have to do monkey-patches to plugin our own passes (see the rmsnorm+quant fusion PR from
@ganyi1996ppo: https://github.com/vllm-project/vllm-ascend/pull/2389). We also want to reuse the existing backend-agnostic FX passes from vllm core, e.g., sequence parallel fusion. But these are all "inductor" passes.

In summary, it would be great if

A HW backend can extend vllm to support custom compiler backend that implements CompilerInterface.
A HW backend with custom compiler backend can reuse existing backend-agnostic FX passes..
Would love to get your inputs.

PS: For "2", we can manually apply the AotDispatch passes and invoke those vllm "inductor" passes from inside the custom compiler backend. Not sure if that would work well though.

Discussion context from:
https://github.com/vllm-project/vllm/pull/23332#issuecomment-3217450860
https://github.com/vllm-project/vllm/pull/23332#issuecomment-3221935847

cc @ProExpertProg @Yikun @wangxiyuan @zou3519 @BoyuanFeng

### Implementation proposal
We propose to add an extra field `oot_compiler` to the `CompilationConfig` for out-of-tree extension to plug in their own piece-wise compiler. Then, `make_compiler` is updated to check and load the `oot_compiler` before instantiating other compilers like inductor.

The changes are listed below:
```diff
diff --git a/vllm/compilation/backends.py b/vllm/compilation/backends.py
index 3361b65a9..a62ce75f8 100644
--- a/vllm/compilation/backends.py
+++ b/vllm/compilation/backends.py
@@ -30,6 +30,8 @@ logger = init_logger(__name__)


 def make_compiler(compilation_config: CompilationConfig) -> CompilerInterface:
+    if compilation.oot_compiler:
+        return resolve_obj_by_qualname(compilation.oot_compiler)()
     if compilation_config.use_inductor:
         if envs.VLLM_USE_STANDALONE_COMPILE and is_torch_equal_or_newer(
                 "2.8.0.dev"):
diff --git a/vllm/config/compilation.py b/vllm/config/compilation.py
index 5c3b22001..90efd89c0 100644
--- a/vllm/config/compilation.py
+++ b/vllm/config/compilation.py
@@ -342,6 +342,8 @@ class CompilationConfig:
         "vllm.linear_attention",
     ]

+    oot_compiler: Optional[str] = None
+
     def compute_hash(self) -> str:
         """
         WARNING: Whenever a new field is added to this config,
```

Below is sample code for how an OOT extension can define its own OOT compiler and plug into vllm. In the OOT compiler, we use aot_autograd to preprocess the graph (do normalization and functionalization etc.) so that we can reuse existing inductor passes or define our own.

```python
...
from torch._dynamo.backends.common import aot_autograd

class TestOotCompiler(CompilerInterface):
    name: str = "TestOotCompiler"

    def compile(
        self,
        graph: fx.GraphModule,
        example_inputs: list[Any],
        compiler_config: dict[str, Any],
        runtime_shape: Optional[int] = None,
        key: Optional[str] = None,
    ) -> tuple[Optional[Callable], Optional[Any]]:
        def compile_inner(fx_graph, inputs):
            # OOT custom fusion pass goes here
            # can reuse existing Inductor passes from vllm core
            return graph

        # AotAutograd
        return aot_autograd(fw_compiler=compile_inner)(graph, example_inputs), None


def test_fusion_with_oot_compiler():
    compilation_config = CompilationConfig(
        level=CompilationLevel.PIECEWISE,
        oot_compiler=TestOotCompiler.__module__ + "." + TestOotCompiler.__name__
    )
    vllm_config = VllmConfig(compilation_config=compilation_config)
    with vllm.config.set_current_vllm_config(vllm_config):
        backend = compilation_config.init_backend(vllm_config)
        model = torch.nn.RMSNorm([2, 3])
        model = torch.compile(
            model,
            fullgraph=True,
            backend=backend)
        input = torch.randn(2, 2, 3)
        model(input)
```

### Alternatives

One may also directly invoke `torch.compile` on the full model with its own backend backend in the OOT extension, such as the practice in TPU (https://github.com/vllm-project/vllm/blob/0235103cbbdb511e6708aae600f759060a797c16/vllm/v1/worker/tpu_model_runner.py#L299). The downside is that we are not able to reuse the piece-wise compilation infrastructure from vllm core.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by ProExpertProg** [comment]
Could you add some more details on the proposed implementation?

**Comment by jgong5** [comment]
> Could you add some more details on the proposed implementation?

Sure. I guess I need to go through the code a bit before coming up with some proposal.

**Comment by jgong5** [comment]
@ProExpertProg Added proposed implementation and alternatives for discussion purpose. Please review and welcome feedbacks. Thanks.

**Comment by jgong5** [comment]
@Yikun @wangxiyuan @whx-sjtu Does it sound good to you? @whx-sjtu are you willing to submit a PR for it?

**Comment by ProExpertProg** [comment]
@jgong5 this looks good to me overall, interested to see how well aot_autograd-generated code works. You might need something like a `FixFunctionalizationPass` to avoid extra copies.

One comment I have, instead of an optional `oot_compiler`, we should just make VllmBackend respect a custom `backend` passed in. This is mentioned in #20283 under the getting rid of `use_inductor` & modifying the `backend` field. The custom backend might need a custom adapter as well because `make_compiler` returns a `CompilerInterface` and not a torch.compile-compatible backend (which is what `backend` is supposed to be), but we can iron out those details later.

Rough changes:
```python
# instead of this:
   if compilation.oot_compiler:
       return resolve_obj_by_qualname(compilation.oot_compiler)()
   if compilation_config.use_inductor:
        if envs.VLLM_USE_STANDALONE_COMPILE and is_torch_equal_or_newer(
                "2.8.0.dev"):
            logger.debug("Using InductorStandaloneAdaptor")
            return InductorStandaloneAdaptor()
        else:
            logger.debug("Using InductorAdaptor")
            return InductorAdaptor()
    else:
        logger.debug("Using EagerAdaptor")
        return EagerAdaptor()
# do this:
   if compilation_config.backend == "inductor":
        if envs.VLLM_USE_STANDALONE_COMPILE and is_torch_equal_or_newer(
                "2.8.0.dev"):
            logger.debug("Using InductorStandaloneAdaptor")
            return InductorStandaloneAdaptor()
        else:
            logger.debug("Using InductorAdaptor")
            return InductorAdaptor()
    elif compilation_config.backend == "eager":
        logger.debug("Using EagerAdaptor")
        return EagerAdaptor()
    else:
        logger.debug(f"Using custom backend: {compilation_config.backend}")
        return resolve_obj_by_qualname(compilation.backend)
```

**Comment by jgong5** [comment]
Thanks @ProExpertProg for the suggestions.

> [@jgong5](https://github.com/jgong5) this looks good to me overall, interested to see how well aot_autograd-generated code works. You might need something like a `FixFunctionalizationPass` to avoid extra copies.
> 

Yes, we can reuse these existing passes. The aot_autograd-generated code is compatible with what we see from inductor's custom passes since both went through the same functionalization/normalization/decomposition passes.

> One comment I have, instead of an optional `oot_compiler`, we should just make VllmBackend respect a custom `backend` passed in. This is mentioned in [#20283](https://github.com/vllm-project/vllm/issues/20283) under the getting rid of `use_inductor` & modifying the `backend` field. The custom backend might need a custom adapter as well because `make_compiler` returns a `CompilerInterface` and not a torch.compile-compatible backend (which is what `backend` is supposed to be), but we can iron out those details later.

My understanding is that the `compilation_config.backend` (which is a torch.compile backend) currently works on the full graph while what we want is a backend that works on the piecewise graphs? Are you suggesting to repurpose this for piecewise graph compilation instead?

**Comment by ProExpertProg** [comment]
> My understanding is that the compilation_config.backend (which is a torch.compile backend) currently works on the full graph while what we want is a backend that works on the piecewise graphs? Are you suggesting to repurpose this for piecewise graph compilation instead?

Please see #20283: currently, `backend` is respected (and passed to torch.compile or Dynamo) if compilation mode is 1 or 2. if compilation mode is 3, `backend` is ignored. We will make it so that with mode 3 (custom `VllmBackend`), `backend` is respected - `backend="inductor"` will be equivalent to `use_inductor=True` and `backend="eager"` will be equivalent to `use_inductor=True` and you can make it so that you can plug your custom `backend` here as well.

**Comment by Yikun** [comment]
@jgong5 @DarkLight1337 @ProExpertProg Thanks all your helps, we also landed it in vllm-ascend yesterday (rmsnorm+quant fusion) https://github.com/vllm-project/vllm-ascend/pull/4168 finally.
